server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -170,6 +201,37 @@ public Index getWriteIndex() {
         return indices.get(indices.size() - 1);
     }
 
+    public Index getWriteIndex(DocWriteRequest<?> request, Metadata metadata) {
+        if (type != Type.TSDB) {
```

Comment:
Maybe a bad idea - could we check that the `getWriteIndex().mode() == TSDB`?

I was sort of hoping that `IndexRouting` could handle this kind of redirection somehow because it "feels right". We already have a specialization of it for tsdb. And, it's, like, testable without building the whole world.

Reply:
Also, maybe it'd be nice not to have to add a type to data streams - they just know their type based on the type of the write index.

Reply:
> Maybe a bad idea - could we check that the getWriteIndex().mode() == TSDB?

👍 

> Also, maybe it'd be nice not to have to add a type to data streams - they just know their type based on the type of the write index.

I will look into whether we can detect how to select to index to write into. I agree this would be nice and cleaner. The `DataStream` class doesn't have access to index settings or other index related metadata, so I need to think on this.